### PR TITLE
Introduce system index types including external

### LINKED ISSUE #67383: External system indices
In the original design of system indices (#50251), every system index would require application specific APIs for access via the REST layer. This works well for features that make use of system indices that live within the Elasticsearch repository but once other pieces of the stack and other features started looking at system indices then there are aspects that cause friction and pain points such as code living in multiple repositories and maintenance/ownership burden. In order to address this, system indices will be split into multiple types. What we've been calling system indices previously, will be known as internal system indices; these indices would belong to code that primarily lives and/or runs within Elasticsearch. Two new types of system indices will be introduced `managed external system indices` and `unmanaged external system indices`. 

Both types of external system indices will make use of existing APIs to access system indices with a special header and value pair. The primary difference will be the management of the index; `managed external system indices` are those that Elasticsearch manages the creation, mappings, and settings of; `unmanaged external system indices` will be completely managed by a component that exists outside of Elasticsearch and performs the management using existing REST APIs.

More details about specific items to be implemented and changed will be added here.

### PR DESCRIPTION
**Pull request by jaymode**
This commit introduces system index types that will be used to
differentiate behavior. Previously system indices were all treated the
same regardless of whether they belonged to Elasticsearch, a stack
component, or one of our solutions. Upon further discussion and
analysis this decision was not in the best interest of the various
teams and instead a new type of system index was needed. These system
indices will be referred to as external system indices. Within external
system indices, an option exists for these indices to be managed by
Elasticsearch or to be managed by the external product.

In order to represent this within Elasticsearch, each system index will
have a type and this type will be used to control behavior.

Closes #67383

There are still some open items that need to be addressed:

- [x] More tests
- [x] ~Figure out integration with security restricted indices~ #69298

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/indices/SystemIndexDescriptor.java [hunk 1]
```diff
@@ -151,6 +175,11 @@ public SystemIndexDescriptor(String indexPattern, String description) {
             Strings.requireNonEmpty(versionMetaKey, "Must supply versionMetaKey if mappings or settings are defined");
             Strings.requireNonEmpty(origin, "Must supply origin if mappings or settings are defined");
         }
+        Objects.requireNonNull(type, "type must not be null");
+        Objects.requireNonNull(allowedElasticProductOrigins, "allowedProductOrigins must not be null");
+        if (type.isInternal() && allowedElasticProductOrigins.isEmpty() == false) {
```
Comment (pugnascotia) [inline server/src/main/java/org/elasticsearch/indices/SystemIndexDescriptor.java:180]: Is the following allowed?

```java
type.isInternal() == false && allowedElasticProductOrigins.isEmpty()
```

File: server/src/main/java/org/elasticsearch/indices/SystemIndices.java [hunk 2]
```diff
@@ -73,6 +78,27 @@ private void checkForDuplicateAliases(Collection<SystemIndexDescriptor> descript
...
+                    map.compute(product, (key, automaton) -> {
+                        Automaton built = SystemIndexDescriptor.buildAutomaton(descriptor.getIndexPattern(), descriptor.getAliasName());
+                        if (automaton != null) {
+                            return Operations.union(automaton, built);
+                        } else {
+                            return built;
+                        }
+                    })
+                )
+            );
```
Comment (pugnascotia) [inline server/src/main/java/org/elasticsearch/indices/SystemIndices.java:95]: What do you think about using a `.flatMap()` here to flatten out the stream processing?

```suggestion
        descriptors.stream()
            .filter(SystemIndexDescriptor::isExternal)
            .flatMap(descriptor -> descriptor.getAllowedElasticProductOrigins().stream())
            .forEach(product ->
                map.compute(product, (key, automaton) -> {
                    Automaton built = SystemIndexDescriptor.buildAutomaton(
                        descriptor.getIndexPattern(), descriptor.getAliasName());
                    return automaton == null ? built : Operations.union(automaton, built);
                })
            );
```

File: server/src/main/java/org/elasticsearch/rest/RestController.java [hunk 3]
```diff
@@ -245,11 +246,19 @@ private void dispatchRequest(RestRequest request, RestChannel channel, RestHandl
...
-                // The ELASTIC_PRODUCT_ORIGIN_HTTP_HEADER indicates that the request is coming from an Elastic product with a plan
-                // to move away from direct access to system indices, and thus deprecation warnings should not be emitted.
+            if (handler.allowSystemIndexAccessByDefault() == false) {
+                // The ELASTIC_PRODUCT_ORIGIN_HTTP_HEADER indicates that the request is coming from an Elastic product and
+                // therefore we should allow a subset of external system index access.
                 // This header is intended for internal use only.
-                client.threadPool().getThreadContext().putHeader(SYSTEM_INDEX_ACCESS_CONTROL_HEADER_KEY, Boolean.FALSE.toString());
+                final String prodOriginValue = request.header(ELASTIC_PRODUCT_ORIGIN_HTTP_HEADER);
+                if (prodOriginValue != null) {
+                    client.threadPool().getThreadContext().putHeader(SYSTEM_INDEX_ACCESS_CONTROL_HEADER_KEY, Boolean.TRUE.toString());
```
Comment (pugnascotia) [inline server/src/main/java/org/elasticsearch/rest/RestController.java:255]: I wondered about capturing `client.threadPool().getThreadContext()` into a local variable - what do you think?

File: x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPlugin.java [hunk 4]
```diff
@@ -237,7 +238,7 @@ protected XPackLicenseState getLicenseState() {
     @Override
     public Collection<SystemIndexDescriptor> getSystemIndexDescriptors(Settings settings) {
         return Collections.singletonList(
-            new SystemIndexDescriptor(ENRICH_INDEX_PATTERN, "Contains data to support enrich ingest processors.")
+            new SystemIndexDescriptor(ENRICH_INDEX_PATTERN, "Contains data to support enrich ingest processors.", Type.INTERNAL, List.of())
```
Comment (pugnascotia) [inline x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPlugin.java:241]: I don't believe you need this change, it's equivalent to the the 2-arg constructor.

**Comment by pugnascotia** [comment]
I see this expression all over the tests:

```
new IndexNameExpressionResolver(new ThreadContext(Settings.EMPTY), new SystemIndices(Map.of()))
```

...and it makes me wonder whether it's worth putting it in a test utility class somewhere. That said, the only value it would add is cutting down some repetition, and at this point I'm not sure it'd be worth the cost of the extra refactoring.

File: test/framework/src/main/java/org/elasticsearch/indices/EmptySystemIndices.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+/*
+ * Licensed to Elasticsearch under one or more contributor
```
Comment (pugnascotia) [inline test/framework/src/main/java/org/elasticsearch/indices/EmptySystemIndices.java:10]: License header problems in this file.

File: test/framework/src/main/java/org/elasticsearch/indices/TestIndexNameExpressionResolver.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+/*
+ * Licensed to Elasticsearch under one or more contributor
```
Comment (pugnascotia) [inline test/framework/src/main/java/org/elasticsearch/indices/TestIndexNameExpressionResolver.java:10]: License header problems in this file.

**Review by pugnascotia — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/indices/SystemIndexDescriptor.java [hunk 7]
```diff
@@ -368,6 +449,16 @@ public Builder setMinimumNodeVersion(Version version) {
             return this;
         }
 
+        public Builder setType(Type type) {
+            this.type = type;
+            return this;
+        }
+
+        public Builder setAllowedStackComponents(List<String> allowedStackComponents) {
```
Comment (gwbrown) [inline server/src/main/java/org/elasticsearch/indices/SystemIndexDescriptor.java:457]: Nit: It's `allowedStackComponents` here in the builder and `allowedElasticProductOrigins` elsewhere, we should probably be consistent.

**Review by AthenaEryma — APPROVED** [review]
Left a few minor comments but overall LGTM

File: server/src/main/java/org/elasticsearch/indices/SystemIndexDescriptor.java [hunk 8]
```diff
@@ -289,13 +334,47 @@ public String checkMinimumNodeVersion(String cause, Version actualMinimumNodeVer
...
-    // TODO: Upgrade handling (reindex script?)
-
     public static Builder builder() {
         return new Builder();
     }
 
+    /**
+     * The specific type of system index that this descriptor represents. System indices have three defined types, which is used to
+     * control behavior. Elasticsearch itself has system indices that are necessary for features of Elasticsearch; these system indices
+     * are referred to as internal system indices. Internal system indices are always managed indices that Elasticsearch itself manages.
```
Comment (williamrandolph) [inline server/src/main/java/org/elasticsearch/indices/SystemIndexDescriptor.java:344]: I think it is worth mentioning that plugins, even third-party plugins, can implement internal indices. A plugin developer trying to implement SystemIndicesPlugin might look here to decide what type of index to choose.

**Review by williamrandolph — APPROVED** [review]
Looks good to me! I have one little thought on a Javadoc.

**Pull request merged**